### PR TITLE
Refactored kubelet's kuberuntime_sandbox

### PR DESCRIPTION
**Pull request by claudiubelu**
#### What type of PR is this?

/kind cleanup
/sig windows

/milestone v1.27

#### What this PR does / why we need it:

Refactors platform specific code into separate files (*_linux.go / *_windows.go / *_others.go)

Adds unit tests for the kuberuntime_sandbox changes.

Co-Authored-By: Claudiu Belu <cbelu@cloudbasesolutions.com>

#### Which issue(s) this PR fixes:

Partial cleanup of issue https://github.com/kubernetes/kubernetes/issues/60338

#### Special notes for your reviewer:

Original PR: https://github.com/kubernetes/kubernetes/pull/109761

This PR contains only the kuberuntime_sandbox changes. Contains a few additional unit tests and cleanups.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by marosset — APPROVED** [review]
/lgtm
Looks good from a Windows standpoint

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox_others.go
Comment (endocrimes) [inline pkg/kubelet/kuberuntime/kuberuntime_sandbox_others.go:27]: I don't _love_ the naming here, but I also can't think of anything meaningfully better off the top of my head.

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox_others_test.go
Comment (endocrimes) [inline pkg/kubelet/kuberuntime/kuberuntime_sandbox_others_test.go:30]: Thanks for adding a test 💯

**Review by endocrimes — COMMENTED** [review]
No behavior change, but is a good stepping stone to making it a bit easier to understand linux-ish vs windows code.

/lgtm

**Comment by endocrimes** [comment]
/skip pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

**Comment by claudiubelu** [comment]
/skip pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox_others.go
Comment (claudiubelu) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox_others.go]: Since this is setting ``SecurityContext`` options in a ``LinuxPodSandboxConfig``, I've renamed this to ``addLinuxSecurityContext``. WDYT?

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:190]: based on https://github.com/kubernetes/enhancements/pull/3275#issuecomment-1121603308 this also should be Linux only. No need to keep it here.

cc @rata

Comment (claudiubelu) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go]: Done

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (rata) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go]: LGTM

**Comment by vinaykul** [comment]
Not sure why the bot is still complaining about pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2 as it appears to have passed after #102884 and the featuregate definition merged. https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/114185/pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2/1631404074370863104

**Comment by vinaykul** [comment]
/skip pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

**Comment by valaparthvi** [comment]
Hi @claudiubelu @kubernetes/sig-node-leads. K8s release v1.27 Bug Triage team here o/
With Code Freeze coming in on March 15, please ensure the PR goes in before that if you would like for it to be a part of v1.27 release. Post Code Freeze, we will only be able to accept Release Blockers and Critical issues.

**Comment by claudiubelu** [comment]
> Hi @claudiubelu @kubernetes/sig-node-leads. K8s release v1.27 Bug Triage team here o/ With Code Freeze coming in on March 15, please ensure the PR goes in before that if you would like for it to be a part of v1.27 release. Post Code Freeze, we will only be able to accept Release Blockers and Critical issues.

It's not a critical issue, I think we can defer this for the next release.

/milestone clear

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox_others.go
Comment (claudiubelu) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox_others.go:1]: ``kuberuntime_sandbox_linux.go`` is only built on ``//go:build linux``, which would exclude other OSes. For example, ``pull-kubernetes-typecheck`` also checks for ``darwin``, so moving this code to ``kuberuntime_sandbox_linux.go`` would cause that job to fail.

Comment (marquiz) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox_others.go:1]: I see that.

Now looking at this It's just a bit odd that we set e.g. selinux options for others than linux. I don't know how risky it is but should we add the "full" `addLinuxSecurityContext` to _linux.go and have a stub or more limited one in this _others.go instead(?)

**Comment by valaparthvi** [comment]
Hello,
It's me again, Parthvi from the Bug Triage Release team 1.28 👋. Since this PR is a spill over from the last milestone, I wanted to check in early this time and make sure it's still on track for 1.28. Please let me know if there's any help you require from my end :)

Thanks.

**Comment by valaparthvi** [comment]
Hello again,

With Code Freeze coming in, on [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496), please make sure this PR goes in by then. We are currently in Week 7, and the code free will be here by Week 9.
Do let me know if I can be of any help :)

**Comment by valaparthvi** [comment]
Hello @claudiubelu,
I am checking in on behalf of the Bug Triage release team. We are only a week away from Code Freeze ([01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496)). Please ensure your PR goes in by then if you are targeting 1.28 release. 

Please let me know if I can be of any assistance :)

**Comment by claudiubelu** [comment]
> Hello @claudiubelu, I am checking in on behalf of the Bug Triage release team. We are only a week away from Code Freeze ([01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496)). Please ensure your PR goes in by then if you are targeting 1.28 release.
> 
> Please let me know if I can be of any assistance :)

Simply waiting for reviews / approval at this point.

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox_others.go
Comment (claudiubelu) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox_others.go:1]: It seems that there is something called SEDarwin: "Security-Enhanced Darwin (SEDarwin) is a port of
access control elements derived from the National Security Agency’s Security Enhanced Linux (SELinux) to Darwin, the Open Source core of Apple’s Mac OSX operating system". I'd expect that most of the SELinux fields we set here to exist for Darwin as well, and this PR just preserves the current behaviour. If they're not, there would be bigger issue here, but that's outside the scope of this PR.

**Comment by furkatgofurov7** [comment]
@claudiubelu hi, Bug Triage Lead here 👋🏼

Since it was communicated in earlier weeks (3 & 1 weeks before), Code Freeze Deadline, this PR has not made to the current milestone, hence I am moving this PR to the next milestone.

/milestone next-candidate

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:186]: is this one linux only? Is there something we configure for Windows for user namespaces?

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:142]: do we need to set linux context on Windows node? Looks counterintuitive

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (SergeyKanzhelev) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:143]: this is far from being obvious in this codebase =). Look at `generatePodSandboxLinuxConfig` - it has many checks for OS version not being Windows

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:141]: I have similar desire to refactor this file when I look at it. This PR is taking a very shallow step in refactoring, which makes it a matter of taste as later comment of @bart0sh indicated.

I would suggest to make a little bit bigger effort of separating clearly Windows-only, Linux-only, and shared settings. So when somebody adds a new feature they will not be confused where they put code for it

Comment (claudiubelu) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:141]: Sure, added a bit more to this PR. Keep in mind that most of this comes from another PR (https://github.com/kubernetes/kubernetes/pull/109761), which was split into 3 (so it's easier to merge, one of them merged, this is still in review: https://github.com/kubernetes/kubernetes/pull/114186), some comments were addressed, a bit extra cleaning.

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (claudiubelu) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:186]: Took a quick glance in ``NamespaceForPod``: it would seem so. First of all, it does have the ``Network`` field, which may be ``runtimeapi.NamespaceMode_NODE`` (host network) or ``runtimeapi.NamespaceMode_POD``, which is relevant for Windows as well. Not sure about the other fields though (``UsernsOptions`` may be relevant, since it's based on the ``Spec``'s ``RuntimeClassName``). I'd leave this here for the time being. But yeah, it's a bit strange that we need ``LinuxPodSandboxConfig`` for Windows.

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (claudiubelu) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:140]: It seems so, according to the other comment I've left.

**Pull request closed without merging**
